camel/utils/python_interpreter.py

```diff
@@ -0,0 +1,359 @@
+            code. All submodules and functions of the modules in the list are
+            importable. Any other import statements will be rejected. :obj:`.`
+            is the seperator between the module and its submodule or its
+            function name. (default: :obj:`[]`)
+    """
+
+    def __init__(self, action_space: Dict[str, Any],
+                 import_white_list: List[str] = []) -> None:
+        self.action_space = action_space
+        self.state = self.action_space.copy()
+        self.fuzz_state = {}
+        self.import_white_list = import_white_list
+
+    def execute(self, code: str, state: Dict[str, Any] = {},
+                fuzz_state: Dict[str,
+                                 Any] = {}, keep_state: bool = True) -> Any:
+        r""" Execute the input python codes in a security environment.
+
+        Args:
+            code (str): Generated python code to be executed.
+            state (Dict[str, Any], optional): External variables that may be
+                used in the generated code. (default: :obj:`{}`)
+            fuzz_state (Dict[str, Any], optional): External varibles that do
+                not have certain varible names. The interpreter will use fuzzy
+                matching to these varibales. For example, if :obj:`fuzz_state`
+                has a variable :obj:`image`, the generated code can use
+                :obj:`input_image` to access it. (default: :obj:`{}`)
+            keep_state (bool, optional):  If :obj:`True`, :obj:`state` and
+                :obj:`fuzz_state` will be kept for later execution. Otherwise,
+                they will be cleared. (default: :obj:`True`)
+
+        Returns:
+            Any: An evaluation value of the last line in the code.
+        """
+        self.state.update(state)
+        self.fuzz_state.update(fuzz_state)
+        try:
+            expression = ast.parse(code)
+        except SyntaxError as e:
+            raise InterpreterError(f"Syntax error in code: {e}")
+
+        result = None
+        for idx, node in enumerate(expression.body):
+            try:
+                line_result = self._execute_ast(node)
+            except InterpreterError as e:
+                if not keep_state:
+                    self.clear_state()
+                msg = (f"Evaluation of the code stopped at line {idx}. "
+                       f"See:\n{e}")
+                # More information can be provided by `ast.unparse()`,
+                # which is new in python 3.9.
+                raise InterpreterError(msg)
+            if line_result is not None:
+                result = line_result
+
+        if not keep_state:
+            self.clear_state()
+
+        return result
+
+    def clear_state(self) -> None:
+        r"""Initialize :obj:`state` and :obj:`fuzz_state`"""
+        self.state = self.action_space.copy()
+        self.fuzz_state = {}
+
+    def _execute_ast(self, expression: ast.AST) -> Any:
```

Comment:
Is this whole file taken from somewhere?

Reply:
The logic of this class was borrowed from hugging face agent. See [python_interpreter.py](https://github.com/huggingface/transformers/blob/8f093fb799246f7dd9104ff44728da0c53a9f67a/src/transformers/tools/python_interpreter.py). However, most of the codes are rewritten to fit our need.

Reply:
@lightaime I think we should refer the original file since it is quite similar

Reply:
Sorry, I'm not very familiar with these licenses and copyright things. As Apache 2.0 request, I should contain original license from HuggingFace team and add our modifications copyright. However, the license checker didn't allow me to do that. Also, I don't know if it is acceptable in our project.

Another solution may be to totally modify its implementation. What do you think?

Reply:
Thank you for your advisions. I have updated the docstring.